test/object-store/sync/flx_sync.cpp

```diff
@@ -299,6 +336,252 @@ TEST_CASE("flx: client reset", "[sync][flx][app][client reset]") {
+            constexpr size_t expected_added_objects = num_objects_added_before - 1;
+            REQUIRE(table->size() == expected_added_objects + num_objects_added_by_harness);
+            size_t count_of_valid_array_data = validate_integrity_of_arrays(table);
+            REQUIRE(count_of_valid_array_data == expected_added_objects);
+        };
+
+        SECTION("Recover: unsuccessful recovery leads to a manual reset") {
+            config_local.sync_config->client_resync_mode = ClientResyncMode::Recover;
+            auto test_reset = reset_utils::make_baas_flx_client_reset(config_local, config_remote, harness.session());
+            test_reset->make_local_changes(std::move(make_local_changes_that_will_fail))
+                ->on_post_reset(std::move(verify_post_reset_state))
+                ->run();
+            auto config_copy = config_local;
+            auto&& [error_future2, err_handler2] = make_error_handler();
+            config_copy.sync_config->error_handler = err_handler2;
+            auto realm_post_reset = Realm::get_shared_realm(config_copy);
+            auto sync_error = std::move(error_future2).get();
+            REQUIRE(before_reset_count == 2);
+            REQUIRE(after_reset_count == 0);
+            REQUIRE(sync_error.error_code == sync::make_error_code(sync::ClientError::auto_client_reset_failure));
+            REQUIRE(sync_error.is_client_reset_requested());
+        }
+
+        SECTION("RecoverOrDiscard: unsuccessful reapply leads to discard") {
+            config_local.sync_config->client_resync_mode = ClientResyncMode::RecoverOrDiscard;
+            auto test_reset = reset_utils::make_baas_flx_client_reset(config_local, config_remote, harness.session());
+            test_reset->make_local_changes(std::move(make_local_changes_that_will_fail))
+                ->on_post_reset(std::move(verify_post_reset_state))
+                ->run();
+
+            auto config_copy = config_local;
+            auto&& [client_reset_future, reset_handler] = make_client_reset_handler();
+            config_copy.sync_config->error_handler = [](std::shared_ptr<SyncSession>, SyncError err) {
+                REALM_ASSERT_EX(!err.is_fatal, err.message);
+            };
+            config_copy.sync_config->notify_after_client_reset = reset_handler;
+            auto realm_post_reset = Realm::get_shared_realm(config_copy);
+            ClientResyncMode mode = client_reset_future.get();
+            REQUIRE(mode == ClientResyncMode::DiscardLocal);
+            realm_post_reset->refresh();
+            auto table = realm_post_reset->read_group().get_table("class_TopLevel");
+            ColKey added = table->get_column_key(added_property_name);
+            REQUIRE(!added);                                        // reverted local changes
+            REQUIRE(table->size() == num_objects_added_by_harness); // discarded all offline local changes
```

Comment:
isn't there a case where a subset of the local changesets could be applied before other changes potentially causing a failure? Could we add a test for that?

Reply:
Yes, this is that test (unless I am not understanding the question here). The recovery does apply several changesets across multiple commits, but then encounters a fatal error. At that point, the state on disk is partially recovered, but the file ident hasn't been updated. Due to the client reset cycle detection, the next time this client connects, it will fallback to Manual mode or DiscardLocal mode. In either case, the partially applied changes are never uploaded.

Reply:
got it